### PR TITLE
Added llama-3 chat template

### PR DESCRIPTION
**Pull request by DifferentialityDevelopment**
This is just simply to add the llama 3 chat template

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -17016,6 +17016,20 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>";
         }
+    } else if (tmpl == "llama3" || (tmpl.find("<|start_header_id|>") != std::string::npos && tmpl.find("<|end_header_id|>") != std::string::npos)) {
+        // Llama 3
+        bool first_message = true; // add BOS at the beginning
+        for (auto message : chat) {
+            std::string role(message->role);
+            if(first_message){
```
Comment (SamuelTallet) [inline llama.cpp:17024]: For consistency
```suggestion
            if (first_message) {
```

**Comment by DifferentialityDevelopment** [comment]
This is my first ever pull request, so please feel free to give me feedback on anything I could improve upon.

File: llama.cpp [hunk 2]
```diff
@@ -17016,6 +17016,20 @@ static int32_t llama_chat_apply_template_internal(
...
+                ss << "<|begin_of_text|>";
+                first_message = false;
+            }
+            ss << "<|start_header_id|>" << message->role << "<|end_header_id|>\n\n" << trim(message->content) << "<|eot_id|>";
```
Comment (SamuelTallet) [inline llama.cpp:17028]: Or more simply
```suggestion
            ss << "<|start_header_id|>" << role << "<|end_header_id|>\n\n" << trim(message->content) << "<|eot_id|>";
```

File: tests/test-chat-template.cpp [hunk 3]
```diff
@@ -73,6 +75,8 @@ int main(void) {
         "SYSTEM: You are a helpful assistant\nUSER: Hello\nASSISTANT: Hi there</s>\nUSER: Who are you\nASSISTANT:    I am an assistant   </s>\nUSER: Another question\nASSISTANT:",
         // CohereForAI/c4ai-command-r-plus
         "<|START_OF_TURN_TOKEN|><|SYSTEM_TOKEN|>You are a helpful assistant<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|USER_TOKEN|>Hello<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>Hi there<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|USER_TOKEN|>Who are you<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>I am an assistant<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|USER_TOKEN|>Another question<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>",
+        // Llama-3
```
Comment (SamuelTallet) [inline tests/test-chat-template.cpp:78]: For consistency
```suggestion
        // Llama 3
```

**Review by SamuelTallet — COMMENTED** [review]
Looks good! I just have some non-critical comments.

File: llama.cpp (hunk 2 above)
Comment (DifferentialityDevelopment) [reply llama.cpp:17028]: Yeah sorry about that, I'm literally not a C++ coder, did this just based on the other examples I could see.
For the most part (day job) I code in C#
Thanks for the feedback!

Comment (SamuelTallet) [reply llama.cpp]: No problem 😉

**Comment by thecivilizedgamer** [comment]
@DifferentialityDevelopment thanks for your quick work on getting a PR open, I pulled your changes to llama.cpp and rebuilt, then tried the new template. I'm seeing some issues with it still, but maybe I'm doing something wrong....? I'm running llama.cpp in server mode, with `--chat-template llama3.` Here's a sample response I got, where it attempts to generate both sides of the conversation:

```
Hello! It's so nice to chat with you again! I'm your friendly virtual assistant. How's your day going so far? Do you need any help or assistance with anything? I'm all ears (or rather, all text) and happy to lend a hand!assistant

Hi! Thanks for checking in! I'm doing alright, just trying to get some work done and then heading out for a bit later. Nothing too exciting. But it's always great to chat with you. How about you? How's your day going?assistant

Thank you! I'm doing well,.......
```

**Comment by DifferentialityDevelopment** [comment]
> @DifferentialityDevelopment thanks for your quick work on getting a PR open, I pulled your changes to llama.cpp and rebuilt, then tried the new template. I'm seeing some issues with it still, but maybe I'm doing something wrong....? I'm running llama.cpp in server mode, with `--chat-template llama3.` Here's a sample response I got, where it attempts to generate both sides of the conversation:
> 
> ```
> Hello! It's so nice to chat with you again! I'm your friendly virtual assistant. How's your day going so far? Do you need any help or assistance with anything? I'm all ears (or rather, all text) and happy to lend a hand!assistant
> 
> Hi! Thanks for checking in! I'm doing alright, just trying to get some work done and then heading out for a bit later. Nothing too exciting. But it's always great to chat with you. How about you? How's your day going?assistant
> 
> Thank you! I'm doing well,.......
> ```

A lot of the GGUF quants had the eot token not being decoded correctly, and subsequently the model output wouldn't stop appropriately, I was seeing the exact same thing in LM Studio, it's not a problem with llama.cpp itself.
For reference this is the GGUF I'm currently using that doesn't have the issue 
https://huggingface.co/lmstudio-community/Meta-Llama-3-8B-Instruct-GGUF

**Comment by x4080** [comment]
Hi, I think you should also modify file utils.hpp
```
    llama_params["stop"].push_back("<|im_end|>"); // chatml
    llama_params["stop"].push_back("<end_of_turn>"); // gemma
    llama_params["stop"].push_back("<|eot_id|>"); // llama 3 <-- need this
```

to make it stop at "<|eot_id|>"

**Comment by thecivilizedgamer** [comment]
> > @DifferentialityDevelopment thanks for your quick work on getting a PR open, I pulled your changes to llama.cpp and rebuilt, then tried the new template. I'm seeing some issues with it still, but maybe I'm doing something wrong....? I'm running llama.cpp in server mode, with `--chat-template llama3.` Here's a sample response I got, where it attempts to generate both sides of the conversation:
> > ```
> > Hello! It's so nice to chat with you again! I'm your friendly virtual assistant. How's your day going so far? Do you need any help or assistance with anything? I'm all ears (or rather, all text) and happy to lend a hand!assistant
> > 
> > Hi! Thanks for checking in! I'm doing alright, just trying to get some work done and then heading out for a bit later. Nothing too exciting. But it's always great to chat with you. How about you? How's your day going?assistant
> > 
> > Thank you! I'm doing well,.......
> > ```
> 
> A lot of the GGUF quants had the eot token not being decoded correctly, and subsequently the model output wouldn't stop appropriately, I was seeing the exact same thing in LM Studio, it's not a problem with llama.cpp itself. For reference this is the GGUF I'm currently using that doesn't have the issue https://huggingface.co/lmstudio-community/Meta-Llama-3-8B-Instruct-GGUF

Ah, that makes sense. I switched models and now it's working perfectly. Thanks to you and everyone else for your efforts :)

**Comment by DifferentialityDevelopment** [comment]
> Hi, I think you should also modify file utils.hpp
> 
> ```
>     llama_params["stop"].push_back("<|im_end|>"); // chatml
>     llama_params["stop"].push_back("<end_of_turn>"); // gemma
>     llama_params["stop"].push_back("<|eot_id|>"); // llama 3 <-- need this
> ```
> 
> to make it stop at "<|eot_id|>"

I'm adding this now, thanks for this!

**Review by ggerganov — APPROVED** [review]

File: llama.cpp [hunk 4]
```diff
@@ -17016,6 +17016,20 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>";
         }
+    } else if (tmpl == "llama3" || (tmpl.find("<|start_header_id|>") != std::string::npos && tmpl.find("<|end_header_id|>") != std::string::npos)) {
+        // Llama 3
+        bool first_message = true; // add BOS at the beginning
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (first_message) {
+                ss << "<|begin_of_text|>";
```
Comment (ngxson) [inline llama.cpp:17025]: You don't need to include `<|begin_of_text|>` if it is already the BOS token (which is the case [here](https://huggingface.co/unsloth/llama-3-8b-Instruct/blob/main/tokenizer_config.json#L2052)). BOS token will be added by tokenizer.

**Review by ngxson — CHANGES_REQUESTED** [review]

File: llama.cpp (hunk 4 above)
Comment (DifferentialityDevelopment) [reply llama.cpp]: Okay thanks! Will do.

**Review by ngxson — APPROVED** [review]
LGTM. Thank you!

**Comment by DifferentialityDevelopment** [comment]
I'm not 100% sure but I think I know why the one test might be failing
tests/test_chat_template.cpp line 79 begins with "<|begin_of_text|>" ie the bos token
though I noticed that none of the other models included their bos token in their expected output, and this lines up with the change @ngxson asked me to make to remove the adding of the bos token to the chat template as it's automatically handled.
So I'm guessing that removing that would make that test pass

**Comment by ggerganov** [comment]
Yes, you need to remove the BOS text from the reference string

**Comment by DifferentialityDevelopment** [comment]
> Yes, you need to remove the BOS text from the reference string

Done!

**Comment by foldl** [comment]
Is this missing? `{% if loop.index0 == 0 %}{% set content = bos_token + content %}`

**Comment by DifferentialityDevelopment** [comment]
> Is this missing? `{% if loop.index0 == 0 %}{% set content = bos_token + content %}`

It is in there yes

`"{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}"`

**Comment by ngxson** [comment]
> Is this missing? `{% if loop.index0 == 0 %}{% set content = bos_token + content %}`

As explained in https://github.com/ggerganov/llama.cpp/pull/6751#discussion_r1571972422 , BOS token is added by tokenizer, so it should not appear in the template

**Comment by ngxson** [comment]
FYI, I added llama3 to list of supported templates in [wiki page](https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template). This PR looks good to me and should get merged now. The failed CI job (build server) doesn't seem to be relevant to changes from this PR. For extra safe, I'll ask to @ggerganov merge it. Thank you all for your efforts.

File: tests/test-chat-template.cpp [hunk 5]
```diff
@@ -46,7 +46,9 @@ int main(void) {
...
+        "{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}"
```
Comment (reneleonhardt) [inline tests/test-chat-template.cpp:51]: Is it not possible to add the comma in this line, so complicated old templates don't have to be duplicated in the diff, but they would be only part of the new added line? Then every new template would be self-contained when inserted.
```suggestion
        ,"{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}"
```

Comment (DifferentialityDevelopment) [reply tests/test-chat-template.cpp:51]: I agree with this

Comment (ngxson) [reply tests/test-chat-template.cpp:51]: In fact, you can also add a trailing comma (at the end of string, not at the beginning) - just like what you did in `expected_output`

File: tests/test-chat-template.cpp [hunk 6]
```diff
@@ -46,7 +46,9 @@ int main(void) {
...
+        // Llama-3
+        ,"{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}"
```
[the remark this answers is not in the document]
Comment (DifferentialityDevelopment) [reply tests/test-chat-template.cpp]: Lol had a feeling that might be the case :)

**Comment by K-Mistele** [comment]
Hi @DifferentialityDevelopment  - per the [model card on hugging face](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct), `<|eot_id|>` is a stop token, but it seems like it might not be the only one? If you look at the pytorch inference code in the model card, you'll see this:

```python
terminators = [
    pipeline.tokenizer.eos_token_id,
    pipeline.tokenizer.convert_tokens_to_ids("<|eot_id|>")
]
```
`<|eot_id|>` is in this PR, but if you check the `tokenizer_config.json` and `tokenizer.json` files in the repo, `pipeline.tokenizer.eos_token_id` in the above snippet refers to `<|end_of_text|>` and is therefore also a stop token. 

While `<|eot_id|>` is in this PR, `<|end_of_text|>` doesn't see to be. Am I missing something?

**Comment by DifferentialityDevelopment** [comment]
> Hi @DifferentialityDevelopment - per the [model card on hugging face](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct), `<|eot_id|>` is a stop token, but it seems like it might not be the only one? If you look at the pytorch inference code in the model card, you'll see this:
> 
> ```python
> terminators = [
>     pipeline.tokenizer.eos_token_id,
>     pipeline.tokenizer.convert_tokens_to_ids("<|eot_id|>")
> ]
> ```
> 
> `<|eot_id|>` is in this PR, but if you check the `tokenizer_config.json` and `tokenizer.json` files in the repo, `pipeline.tokenizer.eos_token_id` in the above snippet refers to `<|end_of_text|>` and is therefore also a stop token.
> 
> While `<|eot_id|>` is in this PR, `<|end_of_text|>` doesn't see to be. Am I missing something?

Nicely spotted! That's actually strange, why would there be two different eot tokens..

**Comment by K-Mistele** [comment]
I think i read somewhere (so take this with a grain of salt, because I don't remember where) that it's an artifact of how they did the instruct-tuning on top of the base model - the base model uses `<|end_of_text|>`, and the instruct model (sometimes?) uses `<|eot_id|>`, and sometimes uses `<|end_of_text|>`.

A cursory glance at the [base model's card](https://huggingface.co/meta-llama/Meta-Llama-3-8B) seems to support this - there's nothing about the custom `terminators` array or using `<|eot_id|>` as a terminator, it seems to use the default tokenizer config which is `<|end_of_text|>`.

Maybe it doesn't need to be added to the chat template, and just needs to be configured as a stop tokens? I'm not sure how that would be done. 

But, the [llama3 github repository](https://github.com/meta-llama/llama3) _does_ [use both as stop tokens](https://github.com/meta-llama/llama3/blob/bb55334adcedfa9f5da66d2e1ed64e6f3dbd82ed/llama/tokenizer.py#L91) in `tokenizer.py`:

```python
def __init__(self, model_path: str):
        """
        Initializes the Tokenizer with a Tiktoken model.

        Args:
            model_path (str): The path to the Tiktoken model file.
        """
        assert os.path.isfile(model_path), model_path

        mergeable_ranks = load_tiktoken_bpe(model_path)
        num_base_tokens = len(mergeable_ranks)
        special_tokens = [
            "<|begin_of_text|>",
            "<|end_of_text|>",
            "<|reserved_special_token_0|>",
            "<|reserved_special_token_1|>",
            "<|reserved_special_token_2|>",
            "<|reserved_special_token_3|>",
            "<|start_header_id|>",
            "<|end_header_id|>",
            "<|reserved_special_token_4|>",
            "<|eot_id|>",  # end of turn
        ] + [
            f"<|reserved_special_token_{i}|>"
            for i in range(5, self.num_reserved_special_tokens - 5)
        ]
        self.special_tokens = {
            token: num_base_tokens + i for i, token in enumerate(special_tokens)
        }
        self.model = tiktoken.Encoding(
            name=Path(model_path).name,
            pat_str=self.pat_str,
            mergeable_ranks=mergeable_ranks,
            special_tokens=self.special_tokens,
        )
        logger.info(f"Reloaded tiktoken model from {model_path}")

        self.n_words: int = self.model.n_vocab
        # BOS / EOS token IDs
        self.bos_id: int = self.special_tokens["<|begin_of_text|>"]
        self.eos_id: int = self.special_tokens["<|end_of_text|>"]
        self.pad_id: int = -1
        self.stop_tokens = {
            self.special_tokens["<|end_of_text|>"],
            self.special_tokens["<|eot_id|>"],
        }
        logger.info(
            f"#words: {self.n_words} - BOS ID: {self.bos_id} - EOS ID: {self.eos_id}"
        )
```

**Comment by DifferentialityDevelopment** [comment]
> I think i read somewhere (so take this with a grain of salt, because I don't remember where) that it's an artifact of how they did the instruct-tuning on top of the base model - the base model uses `<|end_of_text|>`, and the instruct model (sometimes?) uses `<|eot_id|>`, and sometimes uses `<|end_of_text|>`.
> 
> A cursory glance at the [base model's card](https://huggingface.co/meta-llama/Meta-Llama-3-8B) seems to support this - there's nothing about the custom `terminators` array or using `<|eot_id|>` as a terminator, it seems to use the default tokenizer config which is `<|end_of_text|>`.
> 
> Maybe it doesn't need to be added to the chat template, and just needs to be configured as a stop tokens? I'm not sure how that would be done.
> 
> But, the [llama3 github repository](https://github.com/meta-llama/llama3) _does_ [use both as stop tokens](https://github.com/meta-llama/llama3/blob/bb55334adcedfa9f5da66d2e1ed64e6f3dbd82ed/llama/tokenizer.py#L91) in `tokenizer.py`:
> 
> ```python
>         # from the tokenizer init method
>         self.n_words: int = self.model.n_vocab
>         # BOS / EOS token IDs
>         self.bos_id: int = self.special_tokens["<|begin_of_text|>"]
>         self.eos_id: int = self.special_tokens["<|end_of_text|>"]
>         self.pad_id: int = -1
>         self.stop_tokens = {
>             self.special_tokens["<|end_of_text|>"],
>             self.special_tokens["<|eot_id|>"],
>         }
>         logger.info(
>             f"#words: {self.n_words} - BOS ID: {self.bos_id} - EOS ID: {self.eos_id}"
>         )
> ```
> 
> So, maybe it would be best to just mirror that behavior here and add the following?
> 
> ```c++
> llama_params["stop"].push_back("<|eot_id|>");
> 
> // add me too!
> llama_params["stop"].push_back("<|end_of_text|>");
> ```

I've added end_of_text as another stop token for llama 3 in utils.hpp

**Comment by ngxson** [comment]
`<|end_of_text|>` is the EOS token, so you don't need to include it in the list of stop words. In short, server will stop generation if it receives EOS token. For this reason, you don't see `</s>` (for llama2), because it's already the EOS of llama2

**Comment by DifferentialityDevelopment** [comment]
> `<|end_of_text|>` is the EOS token, so you don't need to include it in the list of stop words. In short, server will stop generation if it receives EOS token. For this reason, you don't see `</s>` (for llama2), because it's already the EOS of llama2

Should I remove it then?

**Comment by ngxson** [comment]
Yes, it should be removed. If we decide to add EOS token as stop sequence, we will also need to add for other templates (`</s>`, `<|EOT|>`, ...)

**Comment by teleprint-me** [comment]
`<|eot_id|>` is [End of Turn](https://github.com/meta-llama/llama3/blob/main/llama/tokenizer.py#L70).

**Comment by phymbert** [comment]
@ngxson @ggerganov Can we merge it ?

**Comment by phymbert** [comment]
Yes, I see the other after, better to wait

**Comment by reneleonhardt** [comment]
> Yes it looks good to me. I’m just wondering if we want to wait for the other PR that allows converting the model, then test the converted model with this template before actually merge it?

Will converting the model help fixing the `"broken IQ and imatrix quants"`, or are all problems originating from the base model itself?
https://github.com/ggerganov/llama.cpp/issues/6747#issuecomment-2067643457

**Comment by ggerganov** [comment]
> Yes it looks good to me. I’m just wondering if we want to wait for the other PR that allows converting the model, then test the converted model with this template before actually merge it?

Yes, let's first merge #6745

**Comment by x4080** [comment]
Hi, in the latest version "<|eot_id|>" appears at the end of conversation, it seems utils.hpp dont have the stop token now

**Comment by DifferentialityDevelopment** [comment]
It's due to a different pull request that got merged, if you look at the current utils.hpp you'll see what I mean.
I'll look into it today, I also noticed the same

**Comment by thecivilizedgamer** [comment]
I'm seeing the same issue, with not only Llama 3 but also Phi 3 as described in this issue https://github.com/ggerganov/llama.cpp/issues/6903

Should a new issue be opened specifically for the Llama 3 stop token problem?

**Comment by DifferentialityDevelopment** [comment]
> I'm seeing the same issue, with not only Llama 3 but also Phi 3 as described in this issue #6903
> 
> Should a new issue be opened specifically for the Llama 3 stop token problem?

This pull request was supposed to fix the problem
https://github.com/ggerganov/llama.cpp/pull/6860

**Comment by thecivilizedgamer** [comment]
> > I'm seeing the same issue, with not only Llama 3 but also Phi 3 as described in this issue #6903
> > Should a new issue be opened specifically for the Llama 3 stop token problem?
> 
> This pull request was supposed to fix the problem #6860
> 
> I'm downloading a quant that was recently created to see if it exhibits the same behavior

No you're right, I just pulled the latest commit and rebuilt, and it's now working again (for Llama 3, not for Phi 3) 😃

**Comment by DifferentialityDevelopment** [comment]
The latest quant (for llama 3) I just downloaded does not exhibit this behavior, 
Specifically I tested this https://huggingface.co/MaziyarPanahi/Llama-3-8B-Instruct-64k-GGUF

image

**Comment by DifferentialityDevelopment** [comment]
So I had it wrong, you can now specify stop words from the client side

Without adding stop parameter
image

With the addition of stop parameter
image

**Comment by x4080** [comment]
@DifferentialityDevelopment do you mean adding "<|eot_id|>"  as stop token ? Shouldn't it be automatically stop without adding stop token ?

**Comment by DifferentialityDevelopment** [comment]
> @DifferentialityDevelopment do you mean adding "<|eot_id|>" as stop token ? Shouldn't it be automatically stop without adding stop token ?

I'm not all too sure, it depends on how the gguf was created, what the tokenizer config was at the time etc, for instance llama 3 I tested and it stopped just fine, but with phi I had to add it else it would output the eot token, with the parameter it stops upon encountering any of the specified stop tokens.
It's a new thing that was added.

**Comment by teleprint-me** [comment]
Special tokens are not handled by the general API, but the Special tokens are handled by the templates, or at least they were. I've been really busy this past week, so I haven't had time to review what the specific changes were because they were handled previously during chat.

**Pull request merged**